File: ya/local.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
import difflib
import json
from pathlib import Path
from typing import Callable

from .config import data_home


MAX_TEXT_BYTES = 1024 * 1024
MAX_LIST_ENTRIES = 200
MAX_SEARCH_RESULTS = 100
MAX_DIFF_LINES = 200
AUDIT_LOG_MAX_BYTES = 1024 * 1024
MAX_AUDIT_ARCHIVES = 3


def _audit_log_path(index: int = 0) -> Path:
    suffix = "" if index == 0 else f".{index}"
    return data_home() / f"actions{suffix}.jsonl"


def audit_log_files() -> list[Path]:
    """Return existing audit logs from newest to oldest."""
    return [path for index in range(MAX_AUDIT_ARCHIVES + 1) if (path := _audit_log_path(index)).is_file()]


def audit_log_total_bytes() -> int:
    return sum(path.stat().st_size for path in audit_log_files())


def clear_audit_logs() -> list[Path]:
    """Permanently remove the active audit log and its rotated archives."""
    removed = audit_log_files()
    for path in removed:
        path.unlink()
    return removed


def _tail_complete_lines(data: bytes, limit: int) -> bytes:
    """Keep recent complete JSONL records when rotating a pre-existing large log."""
    if len(data) <= limit:
        return data
    tail = data[-limit:]
    newline = tail.find(b"\n")
    return tail[newline + 1 :] if newline >= 0 else b""


def _rotate_audit_logs() -> None:
    for index in range(MAX_AUDIT_ARCHIVES, 0, -1):
        source = _audit_log_path(index - 1)
        destination = _audit_log_path(index)
        if not source.is_file():
            continue
        if index == MAX_AUDIT_ARCHIVES:
            destination.unlink(missing_ok=True)
        if index == 1 and source.stat().st_size > AUDIT_LOG_MAX_BYTES:
            data = _tail_complete_lines(source.read_bytes(), AUDIT_LOG_MAX_BYTES)
            temporary = destination.with_suffix(destination.suffix + ".tmp")
            temporary.write_bytes(data)
            temporary.replace(destination)
            source.unlink()
        else:
            source.replace(destination)


def append_audit_record(record: dict[str, object]) -> None:
    """Append an audit record while keeping the local audit history bounded."""
    encoded = (json.dumps(record, ensure_ascii=True) + "\n").encode("utf-8")
    active = _audit_log_path()
    active.parent.mkdir(parents=True, exist_ok=True)
    if active.is_file() and active.stat().st_size + len(encoded) > AUDIT_LOG_MAX_BYTES:
        _rotate_audit_logs()
    with active.open("ab") as handle:
        handle.write(encoded)


LOCAL_TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "local_list",
            "description": "List up to 200 entries in a directory inside the authorized workspace.",
            "parameters": {"type": "object", "properties": {"path": {"type": "string"}}},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "local_read",
            "description": "Read a UTF-8 text file up to 1 MiB inside the authorized workspace. Sensitive files are blocked.",
            "parameters": {"type": "object", "properties": {"path": {"type": "string"}}, "required": ["path"]},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "local_search",
            "description": "Search file names and non-sensitive UTF-8 text files inside the authorized workspace.",
            "parameters": {
                "type": "object",
                "properties": {"query": {"type": "string"}, "path": {"type": "string"}},
                "required": ["query"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "local_mkdir",
            "description": "Create one directory inside the workspace. Its parent must already exist. The CLI will ask the user before writing.",
            "parameters": {"type": "object", "properties": {"path": {"type": "string"}}, "required": ["path"]},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "local_write",
            "description": "Create or replace a UTF-8 text file inside the workspace. Its parent must already exist. The CLI will show a diff and ask the user before writing.",
            "parameters": {
                "type": "object",
                "properties": {"path": {"type": "string"}, "content": {"type": "string"}},
                "required": ["path", "content"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "local_move",
            "description": "Move or rename a file or directory inside the workspace without replacing an existing destination. The CLI will ask the user before writing.",
            "parameters": {
                "type": "object",
                "properties": {"source": {"type": "string"}, "destination": {"type": "string"}},
                "required": ["source", "destination"],
            },
        },
    },
]


@dataclass(frozen=True)
class LocalAction:
    operation: str
    paths: tuple[Path, ...]
    summary: str
    diff: str | None = None


@dataclass(frozen=True)
class LocalActivity:
    """Metadata about one local tool operation, safe to show in a GUI."""

    operation: str
    paths: tuple[str, ...]
    status: str


class LocalWorkspace:
    """A deliberately small, workspace-confined filesystem tool set."""

    def __init__(
        self,
        root: Path,
        confirm: Callable[[LocalAction], bool],
        on_activity: Callable[[LocalActivity], None] | None = None,
    ):
        resolved = root.expanduser().resolve(strict=True)
        if not resolved.is_dir():
            raise ValueError(f"Workspace is not a directory: {resolved}")
        self.root = resolved
        self.confirm = confirm
        self.on_activity = on_activity

    def _activity(self, operation: str, paths: tuple[Path, ...], status: str) -> None:
        if self.on_activity is None:
            return
        try:
            self.on_activity(LocalActivity(operation, tuple(self._relative(path) for path in paths), status))
        except Exception:
            # Observers are presentation-only and must never affect the agent.
            pass

    @property
    def tool_handlers(self) -> dict[str, Callable[[dict], str]]:
        return {
            "local_list": self.list,
            "local_read": self.read,
            "local_search": self.search,
            "local_mkdir": self.mkdir,
            "local_write": self.write,
            "local_move": self.move,
        }

    def _resolve(self, value: object) -> Path:
        if not isinstance(value, str) or not value.strip():
            raise ValueError("A non-empty path is required.")
        path = Path(value).expanduser()
        target = (path if path.is_absolute() else self.root / path).resolve(strict=False)
        if target != self.root and self.root not in target.parents:
            raise ValueError("Path must remain inside the authorized workspace.")
        return target

    def _relative(self, path: Path) -> str:
        return str(path.relative_to(self.root)) or "."

    def _is_sensitive(self, path: Path) -> bool:
        lowered = path.name.casefold()
        suffix = path.suffix.casefold()
        if any(part == ".git" for part in path.parts):
            return True
        if lowered == ".env" or lowered.startswith(".env."):
            return True
        if lowered in {"id_rsa", "id_dsa", "id_ecdsa", "id_ed25519", "credentials", "credentials.json"}:
            return True
        return suffix in {".pem", ".key", ".p12", ".pfx", ".kdbx", ".der", ".token"} or "secret" in lowered

    def _read_text(self, path: Path, allow_sensitive: bool = False) -> str:
        if self._is_sensitive(path) and not allow_sensitive:
            raise ValueError("Reading sensitive files is blocked in local mode.")
        if not path.is_file():
            raise ValueError(f"Not a file: {path}")
        if path.stat().st_size > MAX_TEXT_BYTES:
            raise ValueError("Text files larger than 1 MiB cannot be read or replaced.")
        data = path.read_bytes()
        if b"\0" in data:
            raise ValueError("Binary files cannot be read in local mode.")
        try:
            return data.decode("utf-8")
        except UnicodeDecodeError as error:
            raise ValueError("Only UTF-8 text files can be read in local mode.") from error

    def _audit(self, operation: str, paths: tuple[Path, ...], status: str, error: str | None = None) -> None:
        record = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "workspace": str(self.root),
            "operation": operation,
            "paths": [self._relative(path) for path in paths],
            "status": status,
        }
        if error:
            record["error"] = error
        append_audit_record(record)

    def _confirm(self, action: LocalAction) -> bool:
        if self.confirm(action):
            return True
        self._audit(action.operation, action.paths, "denied")
        self._activity(action.operation, action.paths, "denied")
        return False

    def list(self, args: dict) -> str:
        path = self._resolve(args.get("path", "."))
        if not path.is_dir():
            raise ValueError(f"Not a directory: {path}")
        entries = []
        for child in sorted(path.iterdir(), key=lambda item: item.name.casefold())[:MAX_LIST_ENTRIES]:
            kind = "symlink" if child.is_symlink() else "directory" if child.is_dir() else "file"
            entries.append({"path": self._relative(child), "type": kind})
        self._activity("list", (path,), "success")
        return json.dumps({"path": self._relative(path), "entries": entries, "limited": len(entries) == MAX_LIST_ENTRIES})

    def read(self, args: dict) -> str:
        path = self._resolve(args.get("path"))
        content = self._read_text(path)
        self._activity("read", (path,), "success")
        return json.dumps({"path": self._relative(path), "content": content}, ensure_ascii=False)

    def search(self, args: dict) -> str:
        query = args.get("query")
        if not isinstance(query, str) or not query.strip():
            raise ValueError("A non-empty search query is required.")
        root = self._resolve(args.get("path", "."))
        if not root.is_dir():
            raise ValueError(f"Not a directory: {root}")
        needle = query.casefold()
        results: list[dict] = []
        for path in sorted(root.rglob("*"), key=lambda item: str(item).casefold()):
            if len(results) >= MAX_SEARCH_RESULTS:
                break
            if path.is_symlink() or not path.is_file() or self._is_sensitive(path):
                continue
            relative = self._relative(path)
            if needle in path.name.casefold():
                results.append({"path": relative, "match": "filename"})
                if len(results) >= MAX_SEARCH_RESULTS:
                    break
            try:
                text = self._read_text(path)
            except ValueError:
                continue
            for line_number, line in enumerate(text.splitlines(), start=1):
                if needle in line.casefold():
                    results.append({"path": relative, "match": "text", "line": line_number, "text": line[:400]})
                    if len(results) >= MAX_SEARCH_RESULTS:
                        break
        self._activity("search", (root,), "success")
        return json.dumps({"query": query, "results": results, "limited": len(results) == MAX_SEARCH_RESULTS}, ensure_ascii=False)

    def mkdir(self, args: dict) -> str:
        path = self._resolve(args.get("path"))
        action = LocalAction("mkdir", (path,), f"Create directory: {path}")
        try:
            if path.exists():
                raise ValueError(f"Path already exists: {path}")
            if not path.parent.is_dir():
                raise ValueError("Parent directory does not exist; create it first.")
            if not self._confirm(action):
                return json.dumps({"status": "denied", "operation": "mkdir", "path": self._relative(path), "reason": "User declined this action."})
            path.mkdir()
            self._audit("mkdir", (path,), "success")
            self._activity("mkdir", (path,), "success")
            return json.dumps({"status": "ok", "operation": "mkdir", "path": self._relative(path)})
        except Exception as error:
            self._audit("mkdir", (path,), "error", str(error))
            self._activity("mkdir", (path,), "error")
            raise

    def write(self, args: dict) -> str:
        path = self._resolve(args.get("path"))
        content = args.get("content")
        if not isinstance(content, str):
            raise ValueError("Text content is required.")
        try:
            if len(content.encode("utf-8")) > MAX_TEXT_BYTES:
                raise ValueError("Text content larger than 1 MiB cannot be written.")
            if path.exists() and path.is_dir():
                raise ValueError(f"Path is a directory: {path}")
            if not path.parent.is_dir():
                raise ValueError("Parent directory does not exist; create it first.")
            # Existing sensitive files are never sent through local_read/search. The
            # local CLI may still show their replacement diff to its owner.
            previous = self._read_text(path, allow_sensitive=True) if path.exists() else None
            diff = None
            summary = f"Create text file: {path}"
            if previous is not None:
                lines = list(difflib.unified_diff(previous.splitlines(keepends=True), content.splitlines(keepends=True), fromfile=str(path), tofile=str(path)))
                if len(lines) > MAX_DIFF_LINES:
                    lines = lines[:MAX_DIFF_LINES] + ["... diff truncated ...\n"]
                diff = "".join(lines)
                summary = f"Replace text file: {path}"
            action = LocalAction("write", (path,), summary, diff)
            if not self._confirm(action):
                return json.dumps({"status": "denied", "operation": "write", "path": self._relative(path), "reason": "User declined this action."})
            path.write_text(content, encoding="utf-8")
            self._audit("write", (path,), "success")
            self._activity("write", (path,), "success")
            return json.dumps({"status": "ok", "operation": "write", "path": self._relative(path)})
        except Exception as error:
            self._audit("write", (path,), "error", str(error))
            self._activity("write", (path,), "error")
            raise

    def move(self, args: dict) -> str:
        source = self._resolve(args.get("source"))
        destination = self._resolve(args.get("destination"))
        action = LocalAction("move", (source, destination), f"Move: {source} -> {destination}")
        try:
            if source == self.root or not source.exists():
                raise ValueError(f"Source does not exist: {source}")
            if destination.exists():
                raise ValueError(f"Destination already exists: {destination}")
            if not destination.parent.is_dir():
                raise ValueError("Destination parent directory does not exist.")
            if source.is_dir() and source in destination.parents:
                raise ValueError("Cannot move a directory into itself.")
            if not self._confirm(action):
                return json.dumps({"status": "denied", "operation": "move", "path": self._relative(source), "reason": "User declined this action."})
            source.rename(destination)
            self._audit("move", (source, destination), "success")
            self._activity("move", (source, destination), "success")
            return json.dumps({"status": "ok", "operation": "move", "source": self._relative(source), "destination": self._relative(destination)})
        except Exception as error:
            self._audit("move", (source, destination), "error", str(error))
            self._activity("move", (source, destination), "error")
            raise

File: src/local.ts
import {
  appendFileSync,
  existsSync,
  lstatSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  realpathSync,
  renameSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import {
  basename,
  dirname,
  isAbsolute,
  join,
  relative,
  resolve,
  sep,
} from "node:path";
import { createTwoFilesPatch } from "diff";
import {
  assertTextSize as nativeAssertTextSize,
  decodeTextFile as nativeDecodeTextFile,
  isSensitiveFile as nativeIsSensitiveFile,
  localLimits,
  localToolDefinitions,
  tailCompleteLines as nativeTailCompleteLines,
} from "ya-core";
import { dataHome } from "./config";
import type { ToolArguments, ToolDefinition, ToolHandler } from "./types";

/** The limits and the tool contracts live in the Rust core. */
const LIMITS = localLimits();

export const MAX_TEXT_BYTES = LIMITS.maxTextBytes;
export const MAX_LIST_ENTRIES = LIMITS.maxListEntries;
export const MAX_SEARCH_RESULTS = LIMITS.maxSearchResults;
export const MAX_DIFF_LINES = LIMITS.maxDiffLines;
export const MAX_AUDIT_ARCHIVES = LIMITS.maxAuditArchives;
export let AUDIT_LOG_MAX_BYTES = LIMITS.auditLogMaxBytes;

export interface LocalAction {
  operation: "mkdir" | "write" | "move";
  paths: string[];
  summary: string;
  diff?: string;
}

export interface LocalActivity {
  operation: "list" | "read" | "search" | "mkdir" | "write" | "move";
  paths: string[];
  status: "success" | "denied" | "error";
}

export type LocalConfirmation = (action: LocalAction) => boolean | Promise<boolean>;
export type LocalActivityObserver = (activity: LocalActivity) => void;

export const LOCAL_TOOLS: ToolDefinition[] = JSON.parse(localToolDefinitions()) as ToolDefinition[];

export function setAuditLogMaxBytesForTesting(value: number): void {
  AUDIT_LOG_MAX_BYTES = value;
}

export function auditLogPath(index = 0): string {
  return join(dataHome(), `actions${index === 0 ? "" : `.${index}`}.jsonl`);
}

export function auditLogFiles(): string[] {
  const paths: string[] = [];
  for (let index = 0; index <= MAX_AUDIT_ARCHIVES; index += 1) {
    const path = auditLogPath(index);
    if (existsSync(path) && statSync(path).isFile()) paths.push(path);
  }
  return paths;
}

export function auditLogTotalBytes(): number {
  return auditLogFiles().reduce((total, path) => total + statSync(path).size, 0);
}

export function clearAuditLogs(): string[] {
  const removed = auditLogFiles();
  for (const path of removed) unlinkSync(path);
  return removed;
}

function tailCompleteLines(data: Buffer, limit: number): Buffer {
  return nativeTailCompleteLines(data, limit);
}

function rotateAuditLogs(): void {
  for (let index = MAX_AUDIT_ARCHIVES; index >= 1; index -= 1) {
    const source = auditLogPath(index - 1);
    const destination = auditLogPath(index);
    if (!existsSync(source) || !statSync(source).isFile()) continue;
    if (existsSync(destination)) unlinkSync(destination);
    if (index === 1 && statSync(source).size > AUDIT_LOG_MAX_BYTES) {
      const data = tailCompleteLines(readFileSync(source), AUDIT_LOG_MAX_BYTES);
      const temporary = `${destination}.${process.pid}.tmp`;
      writeFileSync(temporary, data);
      renameSync(temporary, destination);
      unlinkSync(source);
    } else {
      renameSync(source, destination);
    }
  }
}

export function appendAuditRecord(record: Record<string, unknown>): void {
  const encoded = Buffer.from(`${JSON.stringify(record)}\n`, "utf8");
  const active = auditLogPath();
  mkdirSync(dirname(active), { recursive: true });
  if (existsSync(active) && statSync(active).size + encoded.byteLength > AUDIT_LOG_MAX_BYTES) rotateAuditLogs();
  appendFileSync(active, encoded);
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function expandHome(value: string): string {
  if (value === "~") return homedir();
  if (value.startsWith("~/") || value.startsWith("~\\")) return join(homedir(), value.slice(2));
  return value;
}

export class LocalWorkspace {
  readonly root: string;
  confirm: LocalConfirmation;
  readonly onActivity?: LocalActivityObserver;

  constructor(root: string, confirm: LocalConfirmation, onActivity?: LocalActivityObserver) {
    const expanded = resolve(expandHome(root));
    if (!existsSync(expanded)) throw new Error(`Workspace does not exist: ${expanded}`);
    const resolved = realpathSync(expanded);
    if (!statSync(resolved).isDirectory()) throw new Error(`Workspace is not a directory: ${resolved}`);
    this.root = resolved;
    this.confirm = confirm;
    this.onActivity = onActivity;
  }

  get toolHandlers(): Record<string, ToolHandler> {
    return {
      local_list: (arguments_) => this.list(arguments_),
      local_read: (arguments_) => this.read(arguments_),
      local_search: (arguments_) => this.search(arguments_),
      local_mkdir: (arguments_) => this.mkdir(arguments_),
      local_write: (arguments_) => this.write(arguments_),
      local_move: (arguments_) => this.move(arguments_),
    };
  }

  private activity(operation: LocalActivity["operation"], paths: string[], status: LocalActivity["status"]): void {
    if (!this.onActivity) return;
    try {
      this.onActivity({ operation, paths: paths.map((path) => this.relativePath(path)), status });
    } catch {
      // Presentation observers cannot affect the agent operation.
    }
  }

  private resolveInput(value: unknown): string {
    if (typeof value !== "string" || !value.trim()) throw new Error("A non-empty path is required.");
    const expanded = expandHome(value);
    const candidate = resolve(isAbsolute(expanded) ? expanded : join(this.root, expanded));
    let ancestor = candidate;
    while (!existsSync(ancestor)) {
      const parent = dirname(ancestor);
      if (parent === ancestor) break;
      ancestor = parent;
    }
    const physicalAncestor = existsSync(ancestor) ? realpathSync(ancestor) : ancestor;
    const physical = resolve(physicalAncestor, relative(ancestor, candidate));
    const fromRoot = relative(this.root, physical);
    if (fromRoot === ".." || fromRoot.startsWith(`..${sep}`) || isAbsolute(fromRoot)) {
      throw new Error("Path must remain inside the authorized workspace.");
    }
    return physical;
  }

  private relativePath(path: string): string {
    return relative(this.root, path) || ".";
  }

  private isSensitive(path: string): boolean {
    // Splitting the path is Node path semantics; the policy itself is in Rust.
    return nativeIsSensitiveFile(basename(path).toLocaleLowerCase("und"), path.split(sep));
  }

  private readText(path: string, allowSensitive = false): string {
    if (this.isSensitive(path) && !allowSensitive) throw new Error("Reading sensitive files is blocked in local mode.");
    if (!existsSync(path) || !statSync(path).isFile()) throw new Error(`Not a file: ${path}`);
    nativeAssertTextSize(statSync(path).size);
    return nativeDecodeTextFile(readFileSync(path));
  }

  private audit(operation: string, paths: string[], status: string, error?: string): void {
    const record: Record<string, unknown> = {
      timestamp: new Date().toISOString(),
      workspace: this.root,
      operation,
      paths: paths.map((path) => this.relativePath(path)),
      status,
    };
    if (error) record.error = error;
    appendAuditRecord(record);
  }

  private async confirmed(action: LocalAction): Promise<boolean> {
    if (await this.confirm(action)) return true;
    this.audit(action.operation, action.paths, "denied");
    this.activity(action.operation, action.paths, "denied");
    return false;
  }

  list(arguments_: ToolArguments): string {
    const path = this.resolveInput(arguments_.path ?? ".");
    if (!existsSync(path) || !statSync(path).isDirectory()) throw new Error(`Not a directory: ${path}`);
    const allEntries = readdirSync(path, { withFileTypes: true }).sort((left, right) =>
      left.name.localeCompare(right.name, undefined, { sensitivity: "base" }),
    );
    const entries = allEntries.slice(0, MAX_LIST_ENTRIES).map((entry) => {
      const child = join(path, entry.name);
      const type = entry.isSymbolicLink() ? "symlink" : entry.isDirectory() ? "directory" : "file";
      return { path: this.relativePath(child), type };
    });
    this.activity("list", [path], "success");
    return JSON.stringify({ path: this.relativePath(path), entries, limited: allEntries.length > MAX_LIST_ENTRIES });
  }

  read(arguments_: ToolArguments): string {
    const path = this.resolveInput(arguments_.path);
    const content = this.readText(path);
    this.activity("read", [path], "success");
    return JSON.stringify({ path: this.relativePath(path), content });
  }

  search(arguments_: ToolArguments): string {
    const query = arguments_.query;
    if (typeof query !== "string" || !query.trim()) throw new Error("A non-empty search query is required.");
    const root = this.resolveInput(arguments_.path ?? ".");
    if (!existsSync(root) || !statSync(root).isDirectory()) throw new Error(`Not a directory: ${root}`);
    const needle = query.toLocaleLowerCase("und");
    const results: Array<Record<string, unknown>> = [];
    const paths = this.walkFiles(root);
    for (const path of paths) {
      if (results.length >= MAX_SEARCH_RESULTS) break;
      const relativePath = this.relativePath(path);
      if (basename(path).toLocaleLowerCase("und").includes(needle)) {
        results.push({ path: relativePath, match: "filename" });
        if (results.length >= MAX_SEARCH_RESULTS) break;
      }
      try {
        const text = this.readText(path);
        const lines = text.split(/\r?\n/u);
        for (let index = 0; index < lines.length; index += 1) {
          const line = lines[index] ?? "";
          if (line.toLocaleLowerCase("und").includes(needle)) {
            results.push({ path: relativePath, match: "text", line: index + 1, text: line.slice(0, 400) });
            if (results.length >= MAX_SEARCH_RESULTS) break;
          }
        }
      } catch {
        // Binary, oversized, and sensitive files are intentionally skipped.
      }
    }
    this.activity("search", [root], "success");
    return JSON.stringify({ query, results, limited: results.length >= MAX_SEARCH_RESULTS });
  }

  private walkFiles(root: string): string[] {
    const files: string[] = [];
    const visit = (directory: string): void => {
      const entries = readdirSync(directory, { withFileTypes: true }).sort((left, right) =>
        left.name.localeCompare(right.name, undefined, { sensitivity: "base" }),
      );
      for (const entry of entries) {
        const path = join(directory, entry.name);
        if (entry.isSymbolicLink() || this.isSensitive(path)) continue;
        if (entry.isDirectory()) visit(path);
        else if (entry.isFile()) files.push(path);
      }
    };
    visit(root);
    return files;
  }

  async mkdir(arguments_: ToolArguments): Promise<string> {
    const path = this.resolveInput(arguments_.path);
    const action: LocalAction = { operation: "mkdir", paths: [path], summary: `Create directory: ${path}` };
    try {
      if (existsSync(path)) throw new Error(`Path already exists: ${path}`);
      if (!existsSync(dirname(path)) || !statSync(dirname(path)).isDirectory()) throw new Error("Parent directory does not exist; create it first.");
      if (!(await this.confirmed(action))) {
        return JSON.stringify({ status: "denied", operation: "mkdir", path: this.relativePath(path), reason: "User declined this action." });
      }
      mkdirSync(path);
      this.audit("mkdir", [path], "success");
      this.activity("mkdir", [path], "success");
      return JSON.stringify({ status: "ok", operation: "mkdir", path: this.relativePath(path) });
    } catch (error) {
      this.audit("mkdir", [path], "error", errorMessage(error));
      this.activity("mkdir", [path], "error");
      throw error;
    }
  }

  async write(arguments_: ToolArguments): Promise<string> {
    const path = this.resolveInput(arguments_.path);
    const content = arguments_.content;
    if (typeof content !== "string") throw new Error("Text content is required.");
    try {
      if (Buffer.byteLength(content, "utf8") > MAX_TEXT_BYTES) throw new Error("Text content larger than 1 MiB cannot be written.");
      if (existsSync(path) && statSync(path).isDirectory()) throw new Error(`Path is a directory: ${path}`);
      if (!existsSync(dirname(path)) || !statSync(dirname(path)).isDirectory()) throw new Error("Parent directory does not exist; create it first.");
      const previous = existsSync(path) ? this.readText(path, true) : undefined;
      const action: LocalAction = previous === undefined
        ? { operation: "write", paths: [path], summary: `Create text file: ${path}` }
        : { operation: "write", paths: [path], summary: `Replace text file: ${path}`, diff: limitedDiff(path, previous, content) };
      if (!(await this.confirmed(action))) {
        return JSON.stringify({ status: "denied", operation: "write", path: this.relativePath(path), reason: "User declined this action." });
      }
      writeFileSync(path, content, "utf8");
      this.audit("write", [path], "success");
      this.activity("write", [path], "success");
      return JSON.stringify({ status: "ok", operation: "write", path: this.relativePath(path) });
    } catch (error) {
      this.audit("write", [path], "error", errorMessage(error));
      this.activity("write", [path], "error");
      throw error;
    }
  }

  async move(arguments_: ToolArguments): Promise<string> {
    const source = this.resolveInput(arguments_.source);
    const destination = this.resolveInput(arguments_.destination);
    const action: LocalAction = { operation: "move", paths: [source, destination], summary: `Move: ${source} -> ${destination}` };
    try {
      if (source === this.root || !existsSync(source)) throw new Error(`Source does not exist: ${source}`);
      if (existsSync(destination)) throw new Error(`Destination already exists: ${destination}`);
      if (!existsSync(dirname(destination)) || !statSync(dirname(destination)).isDirectory()) {
        throw new Error("Destination parent directory does not exist.");
      }
      if (lstatSync(source).isDirectory()) {
        const fromSource = relative(source, destination);
        if (fromSource === "" || (!fromSource.startsWith(`..${sep}`) && fromSource !== ".." && !isAbsolute(fromSource))) {
          throw new Error("Cannot move a directory into itself.");
        }
      }
      if (!(await this.confirmed(action))) {
        return JSON.stringify({ status: "denied", operation: "move", path: this.relativePath(source), reason: "User declined this action." });
      }
      renameSync(source, destination);
      this.audit("move", [source, destination], "success");
      this.activity("move", [source, destination], "success");
      return JSON.stringify({
        status: "ok",
        operation: "move",
        source: this.relativePath(source),
        destination: this.relativePath(destination),
      });
    } catch (error) {
      this.audit("move", [source, destination], "error", errorMessage(error));
      this.activity("move", [source, destination], "error");
      throw error;
    }
  }
}

function limitedDiff(path: string, previous: string, next: string): string {
  const patch = createTwoFilesPatch(path, path, previous, next, "", "", { context: 3 });
  const lines = patch.split(/(?<=\n)/u);
  if (lines.length <= MAX_DIFF_LINES) return patch;
  return `${lines.slice(0, MAX_DIFF_LINES).join("")}... diff truncated ...\n`;
}
